feat: Implement SuperfaceTest

martinalbert opened the pull request:
<!--- Tip: You don't have to remove these comments -->

## Description
This PR adds `SuperfaceTest` class, its methods for easier testing and io functions from CLI.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Provide library for easier testing of capabilities.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- If your changes are only to the internals then make sure any function has its attached documentation updated or added (e.g. descriptive name, doc comment, etc.). If your changes are to user-facing APIs or concepts then make sure README.md is updated accordingly (e.g. command help output). -->
- [ ] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [ ] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

martinalbert committed:
chore: add fixtures

martinalbert committed:
feat: add package.json

martinalbert committed:
feat: implement TestConfig
- add TestConfig interfaces
- add TestConfig class and its methods
- add utility functions

martinalbert committed:
feat: add nockBack playback support

martinalbert committed:
chore: update sdk dependancy

martinalbert committed:
chore: format

martinalbert committed:
chore: update package.json
- add mockttp and @types/debug

martinalbert committed:
fix: update TestConfig
 - remove unnecassary async/await
 - update areCapabilitiesLocal method

martinalbert committed:
test: add tests for TestConfig

martinalbert committed:
docs: add simple readme

martinalbert committed:
fix: resolve path to io module and skip tests

martinalbert committed:
test: refactor tests

martinalbert committed:
feat: update error messages

martinalbert committed:
test: refactor tests

martinalbert committed:
fix: do not use toHaveBeenCalledTimes when not called

martinalbert committed:
feat: add utils and errors modules

martinalbert committed:
feat: add payload types to keep any only in interfaces

martinalbert committed:
test: add tests

martinalbert committed:
chore: move errors module to common

martinalbert committed:
feat: remove unnecassary parts of io module

martinalbert committed:
feat: update test-config according to changes

martinalbert committed:
fix: update tests

martinalbert committed:
chore: update fixtures

martinalbert committed:
chore: move assertIsIOError to errors module

martinalbert committed:
feat: add outputstream and tests

martinalbert committed:
chore: format

martinalbert committed:
feat: set name of error based on passed kind

martinalbert committed:
chore: format

martinalbert committed:
feat: update library according to DX planning
- rename TestConfig to SuperfaceTest
- merge test() and run()
- remove nock.back support
- add check for env variable
- update asserts
- update errors
- add parsing env function

martinalbert committed:
chore: format

martinalbert committed:
wip: test: update tests

martinalbert committed:
chore: lint

Commit:
chore: release v0.0.2-beta.0

martinalbert committed:
feat: clean up code

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

martinalbert committed:
chore: update changelog

Commit:
chore: release v0.0.3-beta.0

lukas-valenta commented on package.json:
```diff
@@ -0,0 +1,49 @@
+    "eslint-plugin-import": "^2.24.0",
+    "eslint-plugin-jest": "^24.4.0",
+    "eslint-plugin-simple-import-sort": "^7.0.0",
+    "jest": "^27.0.6",
```
Shouldn't jest be a runtime dependency?

martinalbert replied:
I’m not using jest in library, it’s only for tests.

lukas-valenta replied:
but the user would need it, right? so peer dependency?

martinalbert replied:
I don't think it needs jest, what if user uses other testing framework?

lukas-valenta commented on src/common/format.ts:
```diff
@@ -0,0 +1,90 @@
+ */
+export function matchWildCard(
+  superfaceEnv: string | undefined,
+  sfConfig: CompleteSuperfaceTestConfig
```
Nitpick: put mandatory parameter first and optional after, with ``?``

lukas-valenta commented on src/common/format.ts:
```diff
@@ -0,0 +1,90 @@
+ * SUPERFACE_LIVE_API="profile*"
+ * SUPERFACE_LIVE_API="scope/profile:provider:usecase*"
+ */
+export function matchWildCard(
```
What is the purpose of the wildcards?

martinalbert replied:
To filter what performs should be recorded based on profile/provider/usecase.

lukas-valenta replied:
i get that, but it seems kinda arbitrary that the * works only at the end

martinalbert replied:
I guess I can extend it to be able to have it at the start as well.

lukas-valenta commented on src/superface-test.interfaces.ts:
```diff
@@ -0,0 +1,49 @@
+  TypedProfile,
+  UseCase,
+} from '@superfaceai/one-sdk';
+import { TypedSuperfaceClient } from '@superfaceai/one-sdk/dist/client/client';
```
Do you need typed interfaces now? If not, remove it pls and i'll add it back when i'm doing typed for real. If yes, update sdk to export it properly

martinalbert replied:
I guess not, station uses typed sdk, but it’s not needed when user can enter just ids of profile and use-cases and not use sdk in tests at all. Btw. `TypedSuperfaceClient` is not exported, only `createTypedClient()`. 😢

lukas-valenta replied:
it's not really meant to be used :shrug:

martinalbert replied:
I'll remove it for now. 👍🏻

martinalbert replied:
Now that I removed it, perform of `UseCase` class require `NonPrimitive` not `unknown`, should I export that properly as well?

lukas-valenta commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,787 @@
+  SuperJson,
+  UseCase,
+} from '@superfaceai/one-sdk';
+import { SDKExecutionError } from '@superfaceai/one-sdk/dist/internal/errors';
```
This should be exported, pls update SDK

martinalbert replied:
Seems like it’s not exported in `v0.0.33` 🤷‍♂️

lukas-valenta replied:
oh i meant update it in sdk so it exports it

martinalbert replied:
right 😅

lukas-valenta reviewed (commented):
Looks good at first glance. I'm not a fan of the excessive mocking and some useless test, let's talk about it sometime (there's no point in removing it now..)

martinalbert commented:
Thanks, I left you few comments there. I should be at the office on thursday so we can talk about it. 👍🏻

martinalbert committed:
fix: review fixes

lukas-valenta commented:
cool, that would be great

martinalbert committed:
fix: review fixes
- update sdk to contain errors
- remove typed interfaces

martinalbert committed:
feat: update DX of library
- change sfConfig in constructor to optional
- change error CapabilitiesNotLocal
- change error RecordingsNotFoundError

Commit:
chore: release v0.0.3-beta.1

martinalbert committed:
feat: remove unnecassary check for nockConfig

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

Commit:
chore: release v0.0.3-beta.2

martinalbert committed:
chore: add new lines to error messages

martinalbert committed:
fix: stop recording before perform throws error

martinalbert committed:
fix: update failing tests

martinalbert committed:
chore: use new sdk with exported PerformError

Commit:
chore: release v0.0.3-beta.3

martinalbert committed:
chore: update typescript and one-sdk

martinalbert committed:
feat: update recording mechanism
- implement hiding credentials
- enable injecting functions for pre/post processing of recordings (preprocessing not implemented yet)
- rename hideReqHeaders field to enableReqheadersRecording

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

Commit:
chore: release v0.0.3-beta.4

martinalbert committed:
fix: add try and catch before writing recording

Commit:
chore: release v0.0.3-beta.5

martinalbert committed:
feat: export type for nock recordings

martinalbert committed:
chore: add ast and update sdk

martinalbert committed:
feat: update function for hiding credentials

martinalbert committed:
chore: downgrade typescript

Commit:
chore: release v0.0.3-beta.6

martinalbert committed:
fix: remove origin from path in recordings

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

martinalbert committed:
feat: update processing functions

Commit:
chore: release v0.0.3-beta.7

TheEdda commented on src/common/format.ts:
```diff
@@ -0,0 +1,90 @@
+  if (profilePayload) {
+    if (profilePayload.endsWith('*')) {
+      const regex = RegExp(
+        `^${profilePayload.substr(0, profilePayload.length - 1)}.`
```
Do we require that `.` char to follow or can `Foo*` also match "Foo"?

martinalbert replied:
Seems like it doesn't need to be there, but originally I wanted to match any other character after payload, so correctly it would be `.*` at the end.

TheEdda commented on src/superface-test.utils.ts:
```diff
@@ -0,0 +1,243 @@
+    }
+  }
+
+  return true;
```
Is this correct? If `profileId === undefined` this returns `true` for any provider, which doesn't seem correct to me. Shouldn't this require `profileId` to not be nullable?

martinalbert replied:
Yes, I'll fix that

TheEdda commented on src/superface-test.utils.ts:
```diff
@@ -0,0 +1,243 @@
+        !(recording.body instanceof RegExp) &&
+        !(recording.body instanceof Buffer) &&
+        !Array.isArray(recording.body) &&
+        !('includes' in recording.body) &&
```
Cursed. Can you please add a comment here so that other poor souls know the pain of detecting an array which is not an array but is an array?

TheEdda commented on src/superface-test.utils.ts:
```diff
@@ -0,0 +1,243 @@
+      const recordingPath = new URL(baseUrl + recording.path);
+
+      if (scheme.in === ApiKeyPlacement.PATH) {
+        const regex = new RegExp(`/${scheme.name}=(.*(?=/))/g`);
```
I don't think this will work correctly. The way apikey with PATH placement is substituted is that an url looks like `/api/v1/something/{myApiKey}/resource` and the post-substitution looks like `/api/v1/something/abc124key/resource`. So to correctly find the secret we either need to find the value of the secret (and know the value of the secret) or the original pre-substitution url (which is in the map) to find the correct place of substitution and manually overwrite it.

martinalbert replied:
Thanks! I can try to remove it based on variables specified in super.json for now. It's up to them to load them before test run.

martinalbert committed:
chore: update sdk dependency

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

Commit:
chore: release v0.0.3-beta.8

martinalbert committed:
chore: update nock dependency

martinalbert committed:
feat: move writing recordings logic to io module

martinalbert committed:
wip: fix: load credentials to pass nock matching

martinalbert committed:
feat: mock sdk components with jest mocks

martinalbert committed:
wip: test: use mocked superface components

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

martinalbert committed:
fix: move writeRecordings to OutputStream module

martinalbert committed:
feat: update SuperfaceTest class and its utils
- add new parameter for method `run()` to enable keeping of original records
- run `activate()` function after loading recordings to handle recording errors correctly
- add warning about sensitive information in recordings
- fix regex in `loadCredentials` function for filtering query parameters

martinalbert committed:
feat: add sdk to peerDeps and devDeps

martinalbert committed:
fix: review fixes

martinalbert committed:
chore: update dependencies

martinalbert committed:
fix: review fixes

martinalbert committed:
fix: update imports

martinalbert committed:
feat: add tests and fix implementation

martinalbert committed:
feat: remove throwing error on empty recording file

martinalbert committed:
chore: update sdk and ast deps

martinalbert committed:
chore: update changelog

martinalbert committed:
feat: enable changing name of recording fixture

Commit:
chore: release v0.0.3-beta.9

martinalbert committed:
chore: update readme

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

martinalbert committed:
fix: tag beta releases in CI

lukas-valenta commented on src/superface-test.interfaces.ts:
```diff
@@ -0,0 +1,69 @@
+};
+
+export interface SuperfaceTestConfig {
+  client?: SuperfaceClient;
```
what is the usecase of passing a client in here?

martinalbert replied:
I guess for now, it's only handy for testing and passing mocked super.json there, but it might be more useful with typed client.

lukas-valenta replied:
and is someone using it? i don't like this

martinalbert replied:
Seems like no one is using it, I'll remove it 👍🏻

lukas-valenta commented:
i don't like the way client is used here, but that's mostly wrong on sdk side and it would be way too much work to get working now properly. i'll just have to refactor this along with SDK

Jakub-Vacek commented on CHANGELOG.md:
```diff
@@ -7,4 +7,36 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- Modules for io, formatting and errors.
+
+### Changed
+- **BREAKING CHANGE**: Rename `before` and `after` processing functions to `afterRecordingLoad` and `beforeRecordingSave`
```
Not sure if you need to mark breaking if you released only Betas?

freaz replied:
And no need for changed, removed, fixed either. All goes as first release.

martinalbert replied:
Make sense, I'll move it all into `Added` section.

Jakub-Vacek commented on src/superface-test.utils.ts:
```diff
@@ -0,0 +1,338 @@
+        );
+      }
+    }
+    // if (scheme.in === ApiKeyPlacement.HEADER) {
```
Do we want to finish this?

martinalbert replied:
Yeah, I'll look into it today 👌🏻

Jakub-Vacek commented on src/superface-test.ts:
```diff
@@ -0,0 +1,355 @@
+   * Checks whether current components in sfConfig
+   * are locally linked in super.json.
+   */
+  private async isMapLocal(): Promise<boolean> {
```
I feel that name of function does not really match its body🤷

Jakub-Vacek commented on src/superface-test.utils.test.ts:
```diff
@@ -0,0 +1,428 @@
+      });
+
+      expect(
+        isProviderLocal('provider', 'profile', mockSuperJson.normalized)
```
According to test it looks like name of fn should be `isMalLocal` of `isProfileProviderLocal`

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,460 @@
+          expect(writeRecordingsSpy).not.toHaveBeenCalled();
+        });
+
+        it.skip('loads credentials that were hidden after recording', async () => {
```
Skip?

Jakub-Vacek commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,460 @@
+
+        mocked(matchWildCard).mockReturnValueOnce(true);
+
+        await expect(superfaceTest.run({ input: {} })).resolves.toMatchObject({
```
If you are comparing two object without fns in them it is better to use `toEqual`. `toMatchObject` is is good for checking if object contains at-least expected object but can be problematic in longterm development.

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+
+<img src="https://github.com/superfaceai/testing-lib/blob/main/docs/LogoGreen.png" alt="superface logo" width="150" height="150">
+
+This library enables easy testing of Superface capabilities with exported class `SuperfaceTest` and its method `run()`.
```
```suggestion
This library enables easy testing of Superface capabilities with `SuperfaceTest` class.
```

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+
+## Usage
+
+To test Superface capabilities, initialize a new `SuperfaceTest` instance and call its method `run()` with corresponding superface config.
```
what is `corresponding superface config`?

martinalbert replied:
It's explained bellow, should I use rather `...with corresponding superface components`?

freaz replied:
I s add m referring to fact that it is not clear what is Superface config, and components are even harder to understand.

martinalbert replied:
Hopefully it's better now

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+
+To test Superface capabilities, initialize a new `SuperfaceTest` instance and call its method `run()` with corresponding superface config.
+
+Superface config should contain `profile`, `provider` and `useCase`. You can enter them either in string format (as ids of corresponding components) or as instances of corresponding components.
```
> or as instances of corresponding components

worth example and describing when it is useful.

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+  profile: 'profile',
+  provider: 'provider',
+  useCase: 'useCase',
+});
```
If I initialize it this way, than I can do just this?

```js
superface.run({ input: { one: 1 } })
```

Or when I should initialize SuperfaceTest with specific profile, provider, useCase?

martinalbert replied:
Yes, those components will be stored and used later in method `run()`. If you enter UseCase instance into method `run()` it will overwrite stored component from constructor until some other `run()` call will specify other UseCase instance. (This can be avoided with `beforeEach()`).

freaz commented on README.md:
````diff
@@ -0,0 +1,185 @@
+);
+```
+
+Given superface config is stored in class and used later in `run()` method. You can also pass in instance of `SuperfaceClient`, but it is not required as it gets initialized inside library if not provided.
````
I don't know to what superface config references to? Is that super.json or the:

```js
   {
     profile: 'profile',
     provider: 'provider',
     useCase: 'useCase',
   },
```

martinalbert replied:
The name is based on interface name `SuperfaceTestConfig` 🤷🏻‍♂️

freaz replied:
But it says Superface Test, which defines it more precisely.

Using Superface to refer to exact configuration isn't enough as we have many configurations available.

I would use test config or all SuperfaceTest instance config.

martinalbert replied:
Alright, I removed 'superface config' term

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+
+### Running
+
+To test your capabilities, you can use method `run()`, which encapsulates nock recording and usecase perform. It expects payload of superface configuration (similar to initializing `SuperfaceTest` class) and input.
```
> To test your capabilities, you can use method `run()`,

There is other way? If not I would remove can from sentence

freaz commented on README.md:
```diff
@@ -0,0 +1,185 @@
+
+## Support
+
+If you need any additional support, have any questions or you just want to talk you can do that through our [documentation page](https://docs.superface.ai).
```
```suggestion
If you need any additional support, have any questions or you just want to talk you can do that through our [support page](https://superface.ai/support).
```

freaz commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,460 @@
+
+  describe('run', () => {
+    describe('when preparing configuration', () => {
+      it('throws when config is not prepared', async () => {
```
Imo better to have it seperated for every testX, so when one of them starts failing it is clear which one.

freaz commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,460 @@
+          })
+        ).resolves.toMatchObject({
+          value: undefined,
+        });
```
This doesn't throw as testcase says

freaz commented on src/superface-test.test.ts:
```diff
@@ -0,0 +1,460 @@
+          expect(recorderSpy).not.toHaveBeenCalled();
+        });
+
+        it('loads fixture if it exists, but contains empty array', async () => {
```
This is testing that nock recording file exists, but has no actual recordings?

martinalbert replied:
Yes, it tests that it doesn't throw when there are no recordings in recording file.

freaz reviewed (commented):
Pause before continuing with tests.

freaz commented on package.json:
```diff
@@ -0,0 +1,53 @@
+    "prepush": "yarn build && yarn test && yarn lint && yarn format"
+  },
+  "devDependencies": {
+    "@superfaceai/one-sdk": "^0.0.41-beta.0",
```
FYI: you have different versions in peer and dev dependencies 😱

martinalbert replied:
I'll use the same version, but should be sdk in peerDeps always the same version as sdk in devDeps when we transform to semver versioning?

freaz replied:
With what version of OneSDK you want to teat it?

I'd say it should be latest version from specified range we claim is supported by testing.

martinalbert replied:
I would keep version in peerDeps compatible with version in devDeps so sdk in peerDeps should not differ in major nor minor version of sdk in devDeps and sdk in devDeps should be latest in specified range as you said. Does it make sense?

martinalbert committed:
fix: review fixes

martinalbert committed:
fix: review fixes

lukas-valenta commented on src/superface-test.utils.ts:
```diff
@@ -0,0 +1,369 @@
+          definition.reqheaders[scheme.name] = HIDDEN_CREDENTIALS_PLACEHOLDER;
+        }
+      } else {
+        if (definition.reqheaders?.['X-API-KEY'] !== undefined) {
```
why have this particular header here?

martinalbert replied:
I refactored it, [now it looks](https://github.com/superfaceai/testing-lib/blob/fedf655d745d52e6d26dad5cc125573ad10e0ad8/src/nock.utils.ts#L38-L49) for loaded environment variable in header values. I also created mutation functions for each case and moved it to `nock.utils.ts`.

Jakub-Vacek commented on README.md:
```diff
@@ -0,0 +1,222 @@
+
+<img src="https://github.com/superfaceai/testing-lib/blob/main/docs/LogoGreen.png" alt="superface logo" width="150" height="150">
+
+This library enables easy testing of Superface capabilities with `SuperfaceTest` class.
```
I would maybe add `local Superface capabilities`🤷

Jakub-Vacek commented on README.md:
```diff
@@ -0,0 +1,222 @@
+
+You can then use this return value to test your capabilities (We recommend you to use jest [snapshot testing](https://jestjs.io/docs/snapshot-testing) as seen in example above).
+
+### Recording
```
Nitpicky: Maybe add some general information about two (life/recorded) possible ways of using this repo to the top of the README.

Jakub-Vacek commented on src/common/format.ts:
```diff
@@ -0,0 +1,89 @@
+  );
+}
+
+function testPayload(payload: string, match: string): boolean {
```
Nitpick: I can't really tell what this function does form it's name. Many `isValidPayload` is better?

freaz commented:
@smizell @martinalbert before releasing 1.0.0 I am for renaming it to `@superfaceai/testing` as the `lib` ins't really needed.

martinalbert committed:
chore: update sdk to 0.0.41-beta.2

Commit:
chore: release v0.0.3-beta.10

martinalbert committed:
fix: add manual check instead of ast functions

Commit:
chore: release v0.0.3-beta.11

martinalbert committed:
fix: review fixes

martinalbert committed:
chore: update sdk to 0.0.41-beta.3

martinalbert committed:
wip: feat: implement hiding integration parameters

martinalbert committed:
fix: update interface

martinalbert committed:
feat: rewrap perform result to Result<unknown, string>

martinalbert committed:
fix: hash input into recording path for each test run

Commit:
chore: release v0.0.3-beta.12

martinalbert committed:
feat: update nock.utils
- refactor nock.utils to easily toggl placeholder and credential
- implement hiding integration params
- remove loading credential and overall manipulation with nock scopes
- rename removeCredential... to replaceCredential...

Commit:
fix: definition path resolution

Jakub-Vacek committed:
Merge pull request #2 from superfaceai/fix/definition-path

fix: definition path resolution

martinalbert committed:
feat: remove hashing library

martinalbert committed:
fix: lint

martinalbert committed:
feat: clean up and proper handling URL

martinalbert committed:
fix: remove replaceAll function

freaz commented:
Noticed readme doesn't mention recording processing functions. Is it intentional?

martinalbert committed:
fix: check for empty array

martinalbert commented:
I'll fill it in 👍🏻

martinalbert committed:
fix: review fixes

martinalbert committed:
chore: update changelog and readme

martinalbert committed:
Merge branch 'feature/implement-hiding-integration-params' of https://github.com/superfaceai/testing-lib into feature/implement-hiding-integration-params

martinalbert committed:
fix: do not mutate payload when credential is empty

martinalbert committed:
Merge branch 'feature/implement-hiding-integration-params' of https://github.com/superfaceai/testing-lib into feature/implement-hiding-integration-params

martinalbert committed:
Merge branch 'feature/implement-test-config' into feature/implement-hiding-integration-params

martinalbert committed:
Merge pull request #3 from superfaceai/feature/implement-hiding-integration-params

feat: Implement hiding integration params

martinalbert committed:
chore: update package version

Commit:
chore: release v0.0.3-beta.14

Jakub-Vacek commented on src/nock.utils.ts:
```diff
@@ -0,0 +1,374 @@
+
+  if (
+    definitionURL.pathname === credential ||
+    definitionURL.pathname.includes(credential)
```
I think includes is true for definitionURL.pathname === credential too.

martinalbert replied:
true 👍🏻

smizell committed:
chore: add known limitations to README

smizell committed:
Clarify when the issue happens

Co-authored-by: Martin Albert <martin.albert.187@gmail.com>

martinalbert committed:
fix: review fixes

martinalbert committed:
Merge branch 'feature/implement-test-config' of https://github.com/superfaceai/testing-lib into feature/implement-test-config

martinalbert committed:
Merge pull request #5 from superfaceai/chore/add-known-limitations

Add known limitations to README

freaz commented on src/superface-test.ts:
```diff
@@ -0,0 +1,393 @@
+    const securityValues = this.sfConfig.provider.configuration.security;
+    const baseUrl = configuration.baseUrl;
+
+    if (!record) {
```
Good practice is to have positive conditions.

martinalbert replied:
I opened new PR #6 for this fix

freaz commented on src/superface-test.ts:
```diff
@@ -0,0 +1,393 @@
+    }
+
+    this.boundProfileProvider =
+      await this.sfConfig.client.cacheBoundProfileProvider(
```
`cacheBoundProfileProvider` is deprecated

lukas-valenta replied:
this is okay, we talked about this and i'll fix it when i remove cacheBoundProfileProvider

freaz replied:
👍 ok thanks
